### PR TITLE
Add @siteimprove/alfa-performance package

### PR DESCRIPTION
**Pull request by kasperisager**
This pull request introduces a new package, @siteimprove/alfa-performance, for working with performance measurements. Its design is based on the [User Timing](https://www.w3.org/TR/user-timing/) API, but is implemented independently of runtime environment (browsers vs. Node.js) and aims to solve some of the shortcomings of the User Timing API:

- **Lack of memory bounds**: With the User Timing API, performance entries are added to and kept in an internal buffer whose size can grow without bound. As such, marks and measurements must be cleared out regularly, or disabled entirely, in long running processes to avoid running out of memory.

- **Everything is globally scoped**: With the User Timing API, any code that wishes to track performance measurements do so via a globally scoped `Performance` instance. While this aids discoverability for consumers that sit outside the main code path, such as Real User Measurement (RUM) collectors, it makes it impossible so do things like collecting performance measurements for a particular call tree only, such as a single audit of a given rule set. 

  This problem also relates to the previous point as a singleton `Performance` instance, and its associated performance entries, can never be garbage collected. With locally scoped `Performance` instances, one would be able to simply drop references to these instances when no longer needed and then have the garbage collector do its job.

The implementation of `Performance` introduced in this pull request solves these problems in the following way:

- No global `Performance` instance is available and specific `Performance` instances must be created through `Performance.of()`:

  ```ts
  const performance = Performance.of<string>();
  ```

  Note that keys are not restricted to the `string` type but can be arbitrarily chosen.

- Entries are never buffered, but are instead shipped straight to listeners. To perform measurements between marks, the marks must be passed explicitly:

  ```ts
  const start = performance.mark("start");
  const end = performance.mark("end");
  
  performance.measure("measurement", start, end);
  ```

### CODE REVIEW

File: packages/alfa-performance/src/performance.ts [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+import * as json from "@siteimprove/alfa-json";
+
+import { now } from "./now";
+
+export class Performance<T>
+  implements
+    AsyncIterable<Performance.Entry<T>>,
+    Serializable<Performance.JSON> {
+  public static of<T>(): Performance<T> {
```
Comment (Jym77) [inline packages/alfa-performance/src/performance.ts:13]: Should `T` default to `string` which sounds like it's going to be a frequent case?
(would simplify calls by removing need to provide the type each time)

File: packages/alfa-performance/src/performance.ts [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    return value instanceof Performance;
+  }
+
+  export type Entry<T> = Mark<T> | Measure<T>;
+
+  export namespace Entry {
+    export type JSON<T> = Mark.JSON<T> | Measure.JSON<T>;
+  }
+
+  export class Mark<T> implements Serializable<Mark.JSON<T>> {
```
Comment (Jym77) [inline packages/alfa-performance/src/performance.ts:116]: Should this (and `Measure`) include a differentiating `type` (or similar) method?

**Review by Jym77 — APPROVED** [review]
Looks good. Some possible convenience improvements.

**Comment by kasperisager** [comment]
Great suggestions, thanks! 🙏

**Review by Jym77 — APPROVED** [review]

**Pull request merged**